### ISSUE TITLE
Status updates to K8ssandraCluster cause stale version errors

### ISSUE THREAD

**Issue by jsanda** [k8ssandra/k8ssandra-operator#71]
#66 added support for updating the `K8ssandraCluster.Status` field. Changes are applied via patches. You will see a lot of errors like this in the operator logs:

`2021-08-07T16:41:04.368Z        ERROR   controller-runtime.manager.controller.k8ssandracluster  Reconciler error        {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test", "namespace": "multi-dc-btzr5d", "error": "Operation cannot be fulfilled on k8ssandraclusters.k8ssandra.io \"test\": the object has been modified; please apply your changes to the latest version and try again"}
github.com/go-logr/zapr.(*zapLogger).Error
        /go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:302
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.UntilWithContext
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:99`

and 

`2021-08-07T16:41:04.415Z        ERROR   controller-runtime.manager.controller.k8ssandracluster  Failed to update status for datacenter  {"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test", "namespace": "multi-dc-btzr5d", "CassandraDatacenter": {"namespace": "multi-dc-btzr5d", "name": "dc1"}, "error": "Operation cannot be fulfilled on k8ssandraclusters.k8ssandra.io \"test\": the object has been modified; please apply your changes to the latest version and try again"}
github.com/go-logr/zapr.(*zapLogger).Error
        /go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132
github.com/k8ssandra/k8ssandra-operator/controllers.(*K8ssandraClusterReconciler).Reconcile
        /workspace/controllers/k8ssandracluster_controller.go:112
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:298
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.UntilWithContext
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:99`

These errors are transitory in that the status updates do eventually get persisted. We can avoid the errors though with some refactoring in k8ssandracluster_controller.go. Rather than making multiple `Patch` calls incrementally,  we could instead iterate through the CassandraDatacenter and Stargate objects, apply whatever changes necessary to `K8ssandraCluster.Status`, and then only after we have finished iterating through those objects make a single `Patch` call.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-786) by [Unito](https://www.unito.io)
┆Fix Versions: k8ssandra-2.0.0-alpha
┆Issue Number: K8SSAND-786
┆Priority: Low

**Comment by jdonenine** [comment]
We talked a bit about if a summary should be provided to provide a "smarter" accounting of rolled up status vs the dump of the underlying status.  Overall it seems we like that approach, but need to do some more research to understand what the underlying statuses provided by cass-operator mean and how we can use those to surface different conditions of the cluster.  For example, what should we surface when one of potentially multiple DCs are down, but others are still available?

**Comment by jdonenine** [comment]
I'm going to carve the summary aspects we talked about into a separate issue that can be covered in addition to resolving the transient state errors described here -- more cover what our long term approach will be here.
